### ISSUE TITLE
Severe JobRunr Exception - Deleting a job that's in process

### ISSUE THREAD

**Issue by dmurphy-ambra** [jobrunr/jobrunr#516]
### Runtime information
- __Timestamp__: 2022-07-12T12:31:40.790Z
- __Location__: BackgroundJobServer efc46e4f-e598-4d47-a966-6ade15455ed7
- __JobRunr Version__: 5.1.4
- __StorageProvider__: org.jobrunr.storage.sql.h2.H2StorageProvider
- __Java Version__: 1.8.0-adoptopenjdk
- __Is running from nested jar__: false

### Background Job Servers
- BackgroundJobServer id: efc46e4f-e598-4d47-a966-6ade15455ed7
(workerPoolSize: 32, pollIntervalInSeconds: 15, firstHeartbeat: 2022-07-12T12:11:08.355Z, lastHeartbeat: 2022-07-12T12:31:36.146Z)

#### Concurrent modified jobs:
Job id: 28ad9534-aec7-3c06-98a5-f3a5ec0bc81d
	Local version: 71; Storage version: 72
	Local state: DELETED (at 2022-07-12T12:30:08.670Z) ← PROCESSING (at 2022-07-12T12:30:08.626Z on BackgroundJobServer efc46e4f-e598-4d47-a966-6ade15455ed7) ← ENQUEUED (at 2022-07-12T12:23:04.050Z)
	Storage state: DELETED (at 2022-07-12T12:29:54.260Z) ← PROCESSING (at 2022-07-12T12:29:53.540Z on BackgroundJobServer efc46e4f-e598-4d47-a966-6ade15455ed7) ← ENQUEUED (at 2022-07-12T12:23:04.050Z)

### Exception
```java
org.jobrunr.SevereJobRunrException: Could not resolve ConcurrentJobModificationException
	at org.jobrunr.server.JobZooKeeper.processJobList(JobZooKeeper.java:233)
	at org.jobrunr.server.JobZooKeeper.updateJobsThatAreBeingProcessed(JobZooKeeper.java:93)
	at org.jobrunr.server.JobZooKeeper.run(JobZooKeeper.java:77)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by cefalo-partha** [comment]
I have the same error. However, in our case the reason is known. Basically, if we have a running job that does not respond properly to "DELETE" requests, we start seeing this error.

For example: lets say I submit a job. The job starts PROCESSING. After a while, I DELETE the job. The job thread is interrupted. But if the job does not respond properly to the `InterruptedException`, we start getting this error. I actually think this is fine.

The problem that I have is that this eventually kills the background job server. And if spring-boot actuator is also enabled, takes down the whole app server since the liveness probes start to fail.

I tried to look into the code. There is this code in `JobZooKeeper`:
```
            if (exceptionCount.getAndIncrement() < 5) {
                LOGGER.warn(JobRunrException.SHOULD_NOT_HAPPEN_MESSAGE + " - Processing will continue.", e);
            } else {
                LOGGER.error("FATAL - JobRunr encountered too many processing exceptions. Shutting down.", shouldNotHappenException(e));
                backgroundJobServer.stop();
            }
```
So, basically, 5 of these errors and the `BackgroundJobServer` will stop. Sometimes even if the PROCESSING thread is stopped within a couple of seconds, we get this error. I think this is because the `JobZooKeeper` runs on a schedule and you can get unlucky.

What do you recommend for this? I can't guarantee that my jobs will be stopped as soon as an `InterruptedException` occurs. In our case, there are some DB code (via mybatis) that can't be interrupted and there is a delay of a few seconds before we actually can check the interrupted status of the thread. For an application that may run for months, 5 of these errors are bound to happen sooner or later.

**Comment by rdehuyss** [comment]
Indeed, @cefalo-partha describes the exact reason. This is a duplicate of #467.

Normally, your job should be interruptible. If it's part of an external library off-course this is not ideal - especially IO operations like MyBatis should be interruptible in my eyes. If you're doing some CPU intensive work yourself, you can check whether the Thread is interrupted using `Thread.currentThread().isInterrupted()`. If that is the case, you should throw an `InterruptedException` or finish the job asap.

This will be solved in JobRunr 6 which will be released in Q3.

**Comment by AlexanderBartash** [comment]
I see a similar issue as well.

**Comment by AlexanderBartash** [comment]
It seems to be that this bug is a bit more serious than it might look initially because jobs get run multiple times which in some cases is worse than not running them at all.

It looks like Jobrunr is updating 'updatedAt' field on all running jobs and if that coincides in time with some job transferring to SUCCEEDED or FAILED state the transactions conflict.

As the result of this bug the job is not being marked as finished and gets run another time, which I do see in logs:
```
[2022-08-31 06:30:02.093] DEBUG o.j.s.AbstractJobScheduler.saveJob:135 - Created Job with id 401cc7c0-52e3-499e-b501-07e736521540
[2022-08-31 06:30:08.379] [backgroundjob-worker-pool-11-thread-3]  DEBUG o.j.s.BackgroundJobServer.processJob:230 - Submitted BackgroundJobPerformer for job 401cc7c0-52e3-499e-b501-07e736521540 to executor service
[2022-08-31 06:30:08.384] [backgroundjob-worker-pool-11-thread-4]  DEBUG o.j.s.BackgroundJobPerformer.updateJobStateToProcessingRunJobFiltersAndReturnIfProcessingCanStart:71 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') processing started

### The first execution
[2022-08-31 06:30:08.385] [backgroundjob-worker-pool-11-thread-4]  TRACE o.j.s.BackgroundJobPerformer.runActualJob:85 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') is running

[2022-08-31 06:35:25.913] [backgroundjob-worker-pool-11-thread-4]  DEBUG o.j.s.BackgroundJobPerformer.updateJobStateToSucceededAndRunJobFilters:99 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') processing succeeded
[2022-08-31 06:35:25.915] [backgroundjob-worker-pool-11-thread-4]  ERROR o.j.s.BackgroundJobPerformer.updateJobStateToSucceededAndRunJobFilters:109 - ERROR - could not update job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') to SUCCEEDED state
org.jobrunr.storage.ConcurrentJobModificationException: The following jobs where concurrently updated: 401cc7c0-52e3-499e-b501-07e736521540
	at org.jobrunr.storage.sql.common.JobTable.save(JobTable.java:77)
	at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.save(DefaultSqlStorageProvider.java:176)
	at org.jobrunr.storage.ThreadSafeStorageProvider.save(ThreadSafeStorageProvider.java:105)
	at org.jobrunr.server.BackgroundJobPerformer.saveAndRunStateRelatedJobFilters(BackgroundJobPerformer.java:139)
	at org.jobrunr.server.BackgroundJobPerformer.updateJobStateToSucceededAndRunJobFilters(BackgroundJobPerformer.java:101)
	at org.jobrunr.server.BackgroundJobPerformer.performJob(BackgroundJobPerformer.java:63)
	at org.jobrunr.server.BackgroundJobPerformer.run(BackgroundJobPerformer.java:41)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
[2022-08-31 06:41:23.484] [backgroundjob-worker-pool-11-thread-1]  DEBUG o.j.s.BackgroundJobServer.processJob:230 - Submitted BackgroundJobPerformer for job 401cc7c0-52e3-499e-b501-07e736521540 to executor service
[2022-08-31 06:41:23.489] [backgroundjob-worker-pool-11-thread-2]  DEBUG o.j.s.BackgroundJobPerformer.updateJobStateToProcessingRunJobFiltersAndReturnIfProcessingCanStart:71 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') processing started


### The second execution
[2022-08-31 06:41:23.489] [backgroundjob-worker-pool-11-thread-2]  TRACE o.j.s.BackgroundJobPerformer.runActualJob:85 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') is running

[2022-08-31 06:45:33.576] [backgroundjob-worker-pool-11-thread-2]  DEBUG o.j.s.BackgroundJobPerformer.updateJobStateToSucceededAndRunJobFilters:99 - Job(id=401cc7c0-52e3-499e-b501-07e736521540, jobName='xxx') processing succeeded
```

**Comment by rdehuyss** [comment]
Hi @AlexanderBartash : let's perhaps connect as I'm not completely following. I've reached out to you via LinkedIn.

**Comment by rdehuyss** [comment]
Could you also attach the automatic report generation from JobRunr? Normally, if such an error occurs, JobRunr allows to generate a Github issue automatically from the dashboard. 

This info helps in diagnosing root causes.

**Comment by rdehuyss** [comment]
We did a lot of improvements related to this and think this is resolved in the latest version of JobRunr (v7.4).

If still encountering this issue, feel free to reopen this issue.
